### PR TITLE
Concept notebooks for composite subset to scope out work

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to address, "Investigate: test existing composite subset support and scope additional work."

- [x] Spatial subsets
- [x] Spectral subsets

### TODO for reviewers

1. Look at the two concept notebooks I added in this PR.
2. Discuss my recommendations in each.
3. Create follow-up tickets as needed.
4. You can close this PR without merge or merge it. I have no preference.
5. Mark https://jira.stsci.edu/browse/JDAT-4573 as Done.

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-4573)

### CODE REVIEW

**Comment by javerbukh** [comment]
Thank you for investigating this so thoroughly! My understanding of the priorities for the points you bring up:

1. The top priority is making it clear to users what `import_region` can and cannot do with some combination of `NotImplemented` errors, documentation, and instructions in the UI.
2. Investigate using compound region for round tripping composite subsets.
3. Create a compound region-like thing for SpectralRegion objects.

I think some of the things you mentioned in the concept notebooks can be a part of those top 3 (for example, only allowing combination_mode = "or" for a filename would be part of 1, same with one file only being able to store one subset). Please let me know if I understood things correctly!

**Comment by pllim** [comment]
(2) Depends on how thorough a roundtrip you really want. If user does not mind the simplified version, then we're basically done for spectral regions. 

(3) Depends on (2).

**Comment by camipacifici** [comment]
Sorry I am late here. I am ok with the recommendations in the imviz concept notebook. Will be looking at the specviz one shortly

**Comment by camipacifici** [comment]
Comments about the specviz notebook, assuming that one day we will be able to save multiple subsets in the same file.
I see a few use cases here:
- user who wants to save many single subsets and reload them at once with combination 'new'. They are all separated anyway.
- user who wants to save many single subsets and reload them at once. They will remember which ones need to be combined together and can do that with a separate method after loading, e.g., load subset 1 and subset 2 and then combine them with 'or' to make a new subset 3.
- user who does not want to remember how they have combined subsets. Currently, they can save a single composite subset out and reload it with combination 'or' and everything will be fine. If they want to save all the subsets together, maybe the csv file could carry the information about the combination? Something like:
subset_name       xmin        xmax      other subset         combination_mode
I see this can be pretty hacky though. Is this what could be fixed by building the equivalent of compound_region in specutils?

Regardless, I think that the default when reading should be 'new', not 'replace'.

**Review by cshanahan1 — COMMENTED** [review]
(Some relevant thoughts to this notebook while working on subset api stuff):

See my comment in https://github.com/spacetelescope/jdaviz/pull/3340. The load_regions and get_regions in subset tools should round trip for composite subsets, and they don't currently.

It sounds like the easiest solution to me to just not support CompoundRegions at all to represent Composite Subsets, and use the current expected input format @javerbukh wrote for Subset Tools.import_regions which is corresponding list of subsets and recognized combination modes. I think this is neater honestly - you can only combine two subsets with CompoundRegion so they become nested when combining many which is more confusing than a list. You won't be able to load any arbitrary CompoundRegion that may have been created outside jdaviz, but you can't anyway unless we figure out how to map the 'operator' to a glue combination mode.

**Comment by pllim** [comment]
Re: https://github.com/spacetelescope/jdaviz/pull/3320#pullrequestreview-2490156969

@cshanahan1 , the complication with your suggestion is the serialization. We would have to roll our own data format if user decides to want to save these spatial composite subsets, send the file to collaborator, and then have the collaborator load the file and get back exactly the same thing. Furthermore, such a specialized file format would be useless outside of Jdaviz (unlike the formats supported natively by `regions`).

**Comment by camipacifici** [comment]
I am ok with this new set of recommendations with two additions:
- we will need methods to recombine subsets via API and UI after importing. I do not see tickets for this, but maybe I missed them?
- have a ticket in specutils to see if there is something that can be done about compound regions.

**Comment by pllim** [comment]
Note to self: Cross check with a list of tickets from @cshanahan1 and then open tickets.

**Comment by pllim** [comment]
@camipacifici 

> we will need methods to recombine subsets via API and UI after importing. I do not see tickets for this, but maybe I missed them?

Is this the ticket? https://jira.stsci.edu/browse/JDAT-4533

> have a ticket in specutils to see if there is something that can be done about compound regions

I opened https://jira.stsci.edu/browse/JDAT-5033

* https://github.com/astropy/specutils/issues/1205

File: notebooks/concepts/imviz_composite_subset_api.ipynb [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    "\n",
+    "However, we could utilize [compound region](https://astropy-regions.readthedocs.io/en/latest/compound.html).\n",
+    "\n",
+    "* Pros: Out of the box support from `regions` for AND, OR (a.k.a. ADD), XOR.\n",
+    "* Cons: No concept of REPLACE and REMOVE. `photutils` does not understand compound apertures, so even if `regions` can load them, they are useless for photometry (not a new problem).\n",
+    "\n",
+    "### Recommendations\n",
+    "\n",
+    "1. Acknowledge that this feature will never be used for aperture photometry.\n",
+    "2. Create follow-up ticket to add support for [compound region](https://astropy-regions.readthedocs.io/en/latest/compound.html).\n",
```
Comment (pllim) [inline notebooks/concepts/imviz_composite_subset_api.ipynb:188]: https://jira.stsci.edu/browse/JDAT-5035

File: notebooks/concepts/imviz_composite_subset_api.ipynb [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    "3. Create follow-up ticket to then investigate REPLACE and REMOVE, after (1) is implemented."
```
Comment (pllim) [inline notebooks/concepts/imviz_composite_subset_api.ipynb:189]: https://jira.stsci.edu/browse/JDAT-5036

File: notebooks/concepts/specviz_composite_subset_api.ipynb [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    "If you want support for arbitrary operators from the same ECSV file, this is not possible\n",
+    "without native operator support in `specutils` (see above).\n",
+    "\n",
+    "The `import_region` API can have \"wild\" behaviors depending on `combination_mode`\n",
+    "when a filename is given. This needs to be reined in.\n",
+    "\n",
+    "### Recommendations\n",
+    "\n",
+    "1. Acknowledge that original inputs might be lost upon saving it back out\n",
+    "   because we simplify the composite spectral regions.\n",
```
Comment (pllim) [inline notebooks/concepts/specviz_composite_subset_api.ipynb:226]: https://jira.stsci.edu/browse/JDAT-5037

File: notebooks/concepts/specviz_composite_subset_api.ipynb [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    "2. Acknowledge that one file can only store one subset.\n",
+    "   (Needs extra pondering if you also want to support `\"new\"` to create one subset\n",
+    "   per file entry.)\n",
+    "4. Only allow `combination_mode=\"or\"` when a filename is given.\n",
+    "   Since default is actually `None` (which `glue` translated to `\"replace\"`,\n",
+    "   maybe have to recode `None` to mean `\"or\"` when input is filename.\n",
+    "   Raise exception for any other value."
```
Comment (pllim) [inline notebooks/concepts/specviz_composite_subset_api.ipynb:233]: https://jira.stsci.edu/browse/JDAT-5038 (modified with Cami's comments)

**Comment by pllim** [comment]
Tickets are opened. Closing without merging.

**Pull request closed without merging**